Mathlib/Analysis/Complex/Hadamard.lean

```diff
@@ -0,0 +1,427 @@
+  · push_neg at h
+    replace h : (0 < sSupNormIm f 0) ∧ (0 < sSupNormIm f 1) := by
+      exact ⟨(lt_of_le_of_ne (sSupNormIm_nonneg f 0) (ne_comm.mp h.1)),
+      (lt_of_le_of_ne (sSupNormIm_nonneg f 1) (ne_comm.mp h.2))⟩
+    exact interpStrip_eq_of_pos f z h.1 h.2
+
+lemma diffContOnCl_interpStrip :
+    DiffContOnCl ℂ (interpStrip f) (verticalStrip 0 1) := by
+  by_cases h : sSupNormIm f 0 = 0 ∨ sSupNormIm f 1 = 0
+  -- Case everywhere 0
+  · eta_expand; simp_rw [interpStrip_eq_of_zero f _ h]; exact diffContOnCl_const
+  -- Case nowhere 0
+  · push_neg at h
+    cases' h with h0 h1
+    rw [ne_comm] at h0 h1
+    apply Differentiable.diffContOnCl
+    intro z
+    eta_expand
+    simp_rw [interpStrip_eq_of_pos f _ (lt_of_le_of_ne (sSupNormIm_nonneg f 0) h0)
+      (lt_of_le_of_ne (sSupNormIm_nonneg f 1) h1)]
+    refine DifferentiableAt.mul ?_ ?_
+    · apply DifferentiableAt.const_cpow (DifferentiableAt.const_sub (differentiableAt_id') 1) _
+      left; simp only [Ne.def, ofReal_eq_zero]; rwa [eq_comm]
+    · refine DifferentiableAt.const_cpow ?_ ?_
+      · apply differentiableAt_id'
+      · left; simp only [Ne.def, ofReal_eq_zero]; rwa [eq_comm]
+
+lemma norm_le_interpStrip_of_mem_verticalClosedStrip_eps (ε : ℝ) (hε : ε > 0) (z : ℂ)
+    (hB : BddAbove ((norm ∘ f) '' (verticalClosedStrip 0 1)))
+    (hd : DiffContOnCl ℂ f (verticalStrip 0 1)) (hz : z ∈ verticalClosedStrip 0 1) :
+    ‖f z‖ ≤  ‖((ε + sSupNormIm f 0) ^ (1-z) * (ε + sSupNormIm f 1) ^ (z) : ℂ)‖ := by
+  simp only [F, abs_invInterpStrip _ _ hε, norm_smul, norm_mul, norm_eq_abs]
+  repeat rw [← ofReal_add]
+  repeat rw [abs_cpow_eq_rpow_re_of_pos (sSupNormIm_eps_pos f hε _) _]
+  simp only [sub_re, one_re]
+  rw [← mul_inv_le_iff, ← one_mul (((ε + sSupNormIm f 1) ^ z.re)), ← mul_inv_le_iff',
+    ← Real.rpow_neg_one, ← Real.rpow_neg_one]
+  · simp only [← Real.rpow_mul (le_of_lt (sSupNormIm_eps_pos f hε _)),
+    mul_neg, mul_one, neg_sub, mul_assoc]
+    simpa [F, abs_invInterpStrip _ _ hε, norm_smul, mul_comm] using
+      norm_mul_invInterpStrip_le_one_of_mem_verticalClosedStrip f ε hε z hd hB hz
+  · simp only [Real.rpow_pos_of_pos (sSupNormIm_eps_pos f hε _) z.re]
+  · simp only [Real.rpow_pos_of_pos (sSupNormIm_eps_pos f hε _) (1-z.re)]
+
+lemma eventuallyle (z : ℂ) (hB : BddAbove ((norm ∘ f) '' (verticalClosedStrip 0 1)))
+    (hd : DiffContOnCl ℂ f (verticalStrip 0 1)) (hz : z ∈ verticalStrip 0 1) :
+    (fun _ : ℝ ↦ ‖f z‖) ≤ᶠ[𝓝[>] 0]
+    (fun ε ↦ ‖((ε + sSupNormIm f 0) ^ (1 - z) * (ε + sSupNormIm f 1) ^ (z) : ℂ)‖ ) := by
+  rw [EventuallyLE, eventually_nhdsWithin_iff, eventually_nhds_iff]
+  use Ioo (-1) 1
+  refine ⟨ ?_, ⟨isOpen_Ioo, by simp only [mem_Ioo, Left.neg_neg_iff, zero_lt_one, and_self]⟩⟩
+  simp only [comp_apply, mem_Ioo, mem_Ioi, and_imp] at *
+  · intro x _ _ hnezero
+    exact norm_le_interpStrip_of_mem_verticalClosedStrip_eps f x (hnezero) z hB hd
+      (mem_of_mem_of_subset hz (preimage_mono Ioo_subset_Icc_self))
```

Comment:
You're trying to flee away from the filters API by rewriting everything to come back to filterless definitions, but you're wrong, filters are fun (and extremely efficient). See: I'm just keeping the last line of your proof, and making the rest disappear.
```suggestion
  filter_upwards [self_mem_nhdsWithin] with ε (hε : 0 < ε) using
    norm_le_interpStrip_of_mem_verticalClosedStrip_eps f ε hε z hB hd
      (mem_of_mem_of_subset hz (preimage_mono Ioo_subset_Icc_self))
```

Reply:
I agree they are super fun! I am just very bad at dealing with them still. I didn't know about the filter_upwards tactic. Would you happen to have a link explaining how it works?